Chore(Docs): AWS EC2 Stress Chaos (#16)
chaosnative/developer-hub-cn: docs/chaos-engineering/chaos-faults/aws/ec2-io-stress.md

- It causes IO Stress chaos on EC2 Instance using an SSM doc for a certain chaos duration.

:::tip Fault execution flow chart
![EC2 IO Stress](./static/images/ec2-io-stress.png)
:::

## Uses

### Uses of the experiment

:::info

- Filesystem read and write is another very common and frequent scenario we find with processes/applications that can result in the impact on its delivery. These problems are generally referred to as "Noisy Neighbour" problems.
- Injecting a rogue process into a target ec2 instance, we starve the main processes/applications (typically pid 1) of the resources allocated to it (where limits are defined) causing slowness in application traffic or in other cases unrestrained use can cause instance to exhaust resources leading to degradation in performance of processes/applications present on the instance. So this category of chaos experiment helps to build the immunity on the application undergoing any such stress scenario.

:::

> ## Prerequisites
> 
> ### Verify the prerequisites
> 
> - Ensure that Kubernetes Version >= 1.17
> 
> **AWS EC2 Access Requirement:**
> 
> - Ensure that SSM agent is installed and running in the target EC2 instance.
> - Ensure to create a Kubernetes secret having the AWS Access Key ID and Secret Access Key credentials in the `CHAOS_NAMESPACE`. A sample secret file looks like:
> 
> ```yaml
> apiVersion: v1
> kind: Secret
> metadata:
>   name: cloud-secret
> type: Opaque
> stringData:
>   cloud_config.yml: |-
>     # Add the cloud AWS credentials respectively
>     [default]
>     aws_access_key_id = XXXXXXXXXXXXXXXXXXX
>     aws_secret_access_key = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
> ```
> 
> - If you change the secret name then please also update the `experiment.yml` ENV values for deriving the respective data from the secret. Also account for the path at which this secret is mounted as a file in the manifest ENV `AWS_SHARED_CREDENTIALS_FILE`.

Reviewer: ```suggestion
## Prerequisites
:::info
- Ensure that Kubernetes Version >= 1.17
- Ensure that SSM agent is installed and running in the target EC2 instance.
- Ensure to create a Kubernetes secret having the AWS Access Key ID and Secret Access Key credentials in the `CHAOS_NAMESPACE`. A sample secret file looks like:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cloud-secret
type: Opaque
stringData:
  cloud_config.yml: |-
    # Add the cloud AWS credentials respectively
    [default]
    aws_access_key_id = XXXXXXXXXXXXXXXXXXX
    aws_secret_access_key = XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
```
- If you change the secret name then please also update the `experiment.yml` ENV values for deriving the respective data from the secret. Also account for the path at which this secret is mounted as a file in the manifest ENV `AWS_SHARED_CREDENTIALS_FILE`.
:::
```

Reviewer: Can we correct the format? The `info` admonition is missing

apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: engine-nginx
spec:
  engineState: "active"
  annotationCheck: "false"
  chaosServiceAccount: ec2-io-stress-sa
  experiments:
  - name: ec2-io-stress
    spec:
      components:
        env:
        - name: VOLUME_MOUNT_PATH
          VALUE: '/tmp'
        # id of the ec2 instance
        - name: EC2_INSTANCE_ID
          value: 'instance-1'
        # region for the ec2 instance
        - name: REGION
          value: 'us-east-1'
```

### MULTIPLE EC2 INSTANCES

Multiple EC2 instances can be targeted in one chaos run. It can be tuned via `EC2_INSTANCE_ID` ENV.

Use the following example to tune this:

[embedmd]:# (./static/manifests/ec2-io-stress/multiple-instances.yaml yaml)
```yaml
# mutilple instance targets
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: engine-nginx
spec:
  engineState: "active"
  annotationCheck: "false"
  chaosServiceAccount: ec2-io-stress-sa
  experiments:
  - name: ec2-io-stress
    spec:
      components:
        env:
        - name: MEMORY_CONSUMPTION
          VALUE: '1024'
        # ids of the ec2 instances
        - name: EC2_INSTANCE_ID

>           value: 'instance-1'

Reviewer 2: ```suggestion
          value: 'instance-1,instance-2'
```

Author: Have provided multiple values in the multiple-instances.yaml example.
